`[cloudformation/custom_resource]` improve output, parametrize, deletion policy

charlesg3 opened the pull request:
### Description of changes
The outputs from custom resources cannot be nested, therefore in order to be able to capture the nested properties of a cluster (e.g. `{"headNode": {"privateIpAddress": "1.2.3.4"}}`) from the custom resource to expose as an output on the CloudFormation stack, we need to flatten the properties.
* Flatten cluster properties, extract the properties from the "cluster" property obtained from the `cluster_create` and `cluster_delete` operations. Propagate relevant information in the stack events on failure.
* Parameterize the bucket, service token and subnet ids to ease in testing and assist in E2E tests in our CI pipeline.
* Capture the head node IP address and validation messages in the outputs of the test stack and validate it in the tests.
* Unify the description on our CloudFormation stacks (see screenshot)
* Add support for a `DeletionPolicy` parameter as well as tests.
* Add a 1-click CFN template for E2E testing and documentation purposes.
* Add tags to lambda function for tracking.

### Tests

The local tests were run against all commits included as of https://github.com/aws/aws-parallelcluster/commit/e8f8bb3c65a5c3b606c4b411df7775d791ae992d

```.sh
$ pytest -m local -s -x ./test_cluster_custom_resource.py
======================================== test session starts =========================================
platform darwin -- Python 3.9.6, pytest-7.2.1, pluggy-1.0.0
rootdir: /Users/cgruenwa/src/cgpc/aws-parallelcluster/cloudformation/tests, configfile: pytest.ini
plugins: xdist-3.1.0, html-3.2.0, mock-3.10.0, typeguard-2.13.3, cov-4.0.0, metadata-2.0.4, datadir-1.
4.1
collected 11 items

test_cluster_custom_resource.py ...........

========================================== warnings summary ==========================================
[...]
============================ 11 passed, 40 warnings in 3870.00s (1:04:29) ============================
```

- Internal docs have been updated to demonstrate this functionality.

- The 1-click CFN has been tested by the following:
  -  Setting the `Bucket` value in the mappings of `cluster-1-click.yaml`
  - Uploading the artifacts to the specified bucket in the appropriate paths
  - Running the script below

```.sh
dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"

random-string()
{

if [[ $OSTYPE == 'darwin'* ]]; then
    echo $RANDOM | md5 | head -c 5; echo;
else
    echo $RANDOM | md5sum | head -c 5; echo;
fi
}

cfn_template=${dir}/cluster-1-click.yaml
stack_name=$(echo "pc-1-click"-`random-string`)
echo "Deploying: " ${cfn_template} "->" ${stack_name}


aws cloudformation deploy \
    --stack-name ${stack_name} \
    --parameter-overrides KeyName=${KEYNAME} \
    --template-file ${cfn_template} \
    --capabilities CAPABILITY_NAMED_IAM CAPABILITY_AUTO_EXPAND

aws cloudformation describe-stacks --stack-name ${stack_name}

aws cloudformation delete-stack --stack-name ${stack_name}
```

### Output:
```
$ ./deploy.sh
Deploying:  [...]/aws-parallelcluster/cloudformation/custom_resource/cluster-1-cli
ck.yaml -> pc-1-click-e40bd

Waiting for changeset to be created..
Waiting for stack create/update to complete
Successfully created/updated stack - pc-1-click-e40bd
{
    "Stacks": [
        {
            "StackId": "arn:[...]",
            "StackName": "pc-1-click-e40bd",
            "ChangeSetId": "[...]",
            "Description": "AWS ParallelCluster CloudFormation Template\n",
            "Parameters": [
                {
                    "ParameterKey": "KeyName",
                    "ParameterValue": "REDACTED"
                },
                {
                    "ParameterKey": "AvailabilityZone",
                    "ParameterValue": "us-east-2a"
                }
            ],
            "CreationTime": "2023-03-20T01:00:25.278000+00:00",
            "LastUpdatedTime": "2023-03-20T01:00:30.832000+00:00",
            "RollbackConfiguration": {},
            "StackStatus"
```

----

![image](https://user-images.githubusercontent.com/6087509/226223472-e571c8c7-6f9f-4998-86e6-d9ad88f10442.png)


----

Validating tags:

![image](https://user-images.githubusercontent.com/6087509/226402482-9f81cf60-e393-4c13-8424-2ceacd68f22b.png)


### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

charlesg3 committed:
`[cloudformation/custom_resource]` Improve output handling and validation.

charlesg3 committed:
`[cloudformation/custom_resource]` Improve output handling and validation.

charlesg3 committed:
`[cloudformation/custom_resource]` Parametrize template and testing infrastructure.

charlesg3 committed:
`[cloudformation/custom_resource]` Add support for deletion policy.

charlesg3 committed:
`[cloudformation/custom_resource]` Unify format of template descriptions.

charlesg3 committed:
`[cloudformation/custom_resource]` Unify format of template descriptions.

charlesg3 committed:
`[cloudformation/custom_resource]` Add 1-click cluster for documentation and e2e testing.

eantonin commented on cloudformation/custom_resource/cluster-1-click.yaml:
```diff
@@ -0,0 +1,98 @@
+          -
+            AmiSearchFilters:
+              Owner: self
+          - !Ref AWS::NoValue
```
"silly" question, non blocking: have you checked an "empty" DevSettings section results in a valid YAML?

charlesg3 replied:
I believe the NoValue here makes it not include the DevSettings, however I have checked (during debugging) that if this isn't included the YAML is still valid (but it will fail to find an official AMI)

eantonin commented on cloudformation/custom_resource/cluster.yaml:
```diff
@@ -208,6 +210,13 @@ Resources:
+
+              deletion_policy = properties.get("DeletionPolicy", "Delete")
+              if deletion_policy not in {"Retain", "Delete"}:
+                  raise ValueError("DeleetionPolicy must be one of [\"Retain\", \"Delete\"].")
```
nitpick, non-blocking: typo

charlesg3 replied:
thanks -- fixed.

eantonin replied:
nitpick, non-blocking: Still see `Deleetion`

eantonin commented on cloudformation/custom_resource/cluster.yaml:
```diff
@@ -14,6 +14,12 @@ Parameters:
     Type: CommaDelimitedList
     Default: ''
 
+  CustomBucket:
```
question, non blocking: Why don't we put this in `Mappings` as we did for the 1-click? So it won't be directly visible to customers via the Console (if I'm not mistaken)

charlesg3 replied:
The cluster yaml is something that we will upload and test from the uploaded artifact -- thus we want it to be immutable so testing it as-is is a win here.

For the 1-click -- we can test that (single) file locally while pulling the other artifacts transitively (overwriting the custom bucket parameter from the mapping), therefore it is much easier to say mutate it with a sed expression and test the mutated version, while validating the other templates that have been uploaded.

eantonin commented on cloudformation/tests/conftest.py:
```diff
@@ -57,6 +64,30 @@ def random_stack_name_fixture():
+
+
+@pytest.fixture(scope="session")
+def bucket(pytestconfig):
```
nitpick, non-blocking, just wondering why you did not use the convention of calling the method `XXX_fixture `and then assigning the `name`, as you do below and in other parts of the code

charlesg3 replied:
Good call -- it's because the bucket fixture isn't use in this file so it doesn't create a shadow but this is more consistent. Fixed.

eantonin commented on cloudformation/tests/conftest.py:
```diff
@@ -8,6 +9,12 @@
 PUBPRIV_TEMPLATE = "../networking/public-private.cfn.json"
 
 
+def pytest_addoption(parser):
```
question, non-blocking. I may be wrong, but it seems to me we are not using this

charlesg3 replied:
This is to provide a convenient mechanism to override some of the settings using pytest parameters.

eantonin commented on cloudformation/custom_resource/cluster.yaml:
```diff
@@ -126,37 +129,61 @@ Resources:
+          def flatten(obj, path=""):
+              """Flatten a nested map using dot-notation for keys."""
+              _dict = obj if isinstance(obj, dict) else {str(i): v for i, v in enumerate(obj)}
+              ret = {path + str(k): v for k, v in _dict.items() if not isinstance(v, (dict, list))}
```
nitpick, non blocking: can we have better names here? this is not very readable

eantonin commented on cloudformation/custom_resource/cluster.yaml:
```diff
@@ -126,37 +129,61 @@ Resources:
-                  kwargs = {**{pcluster.utils.to_snake_case(k): serialize(v) for k, v in properties.items() if k not in {"ServiceToken"}}, "wait": False}
-                  func = {"CREATE": pc.create_cluster, "UPDATE": pc.update_cluster}[event["RequestType"].upper()]
+                  meta_keys = {"ServiceToken"}
+                  kwargs = {**{pcluster.utils.to_snake_case(k): serialize(v) for k, v in properties.items() if k not in meta_keys}, "wait": False}
```
nitpick, non-blocking: cannot you use `drop_keys` here?

charlesg3 replied:
Yep -- done.

eantonin commented on cloudformation/tests/test_cluster.yaml:
```diff
@@ -15,7 +15,7 @@ Parameters:
     Description: Maximum number of compute instances
     Type: Number
-    Default: 10
+    Default: 16
```
nitpick, is this a typo?

charlesg3 replied:
No it is to separate this work from this bug: https://github.com/aws/aws-parallelcluster/pull/5094

charlesg3 committed:
`[cloudformation/custom_resource]` Tag lambda for tracking purposes.

charlesg3 commented:
Thanks for the feedback -- I addressed many of the nits and re-tested with the following script:

```
#!/bin/bash
dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
if [[ $OSTYPE == 'darwin'* ]]; then
    work_dir=`mktemp -d -t pc`
else
    work_dir=`mktemp -d -p "$dir"`
fi

random-string()
{
if [[ $OSTYPE == 'darwin'* ]]; then
    echo $RANDOM | md5 | head -c 5; echo;
else
    echo $RANDOM | md5sum | head -c 5; echo;
fi
}

function cleanup {
  rm -rf ${work_dir}
  echo "Deleted temp working directory $work_dir"
}

# register the cleanup function to be called on the EXIT signal
trap cleanup EXIT

cfn_template=${dir}/cluster-1-click.yaml
stack_name=$(echo "pc-1-click"-`random-string`)

pushd ${work_dir}

local_template=template.yaml
echo "Deploying: " ${work_dir}/${local_template} "->" ${stack_name}
cat ${cfn_template} | sed -e "s/Bucket: .* debug purposes .*/Bucket: '${BUCKET}'/" > ${local_template}

aws cloudformation deploy \
    --stack-name ${stack_name} \
    --parameter-overrides KeyName=${KEYNAME} \
    --template-file ${local_template} \
    --capabilities CAPABILITY_NAMED_IAM CAPABILITY_AUTO_EXPAND

aws cloudformation describe-stacks --stack-name ${stack_name}

aws cloudformation delete-stack --stack-name ${stack_name}

popd
```

->

```
Deploying:  [...]cloudformation/custom_resource/tmp.xwrgIqPB0g/template.yaml -> pc-1-click-c9c0a

Waiting for changeset to be created..
Waiting for stack create/update to complete
Successfully created/updated stack - pc-1-click-c9c0a
{
    "Stacks": [
        {
            "StackId": "arn:aws:cloudformation:us-east-2:00000000000:stack/pc-1-click-c9c0a/[...]",
            "StackName": "pc-1-click-c9c0a",
            "ChangeSetId": "...",
            "Description": "AWS ParallelCluster CloudFormation Template",
            "Parameters": [
                {
                    "ParameterKey": "KeyName",
                    "ParameterValue": "[REDACTED]"
                },
                {
                    "ParameterKey": "AvailabilityZone",
                    "ParameterValue": "us-east-2a"
                }
            ],
            "CreationTime": "2023-03-20T16:11:02.885Z",
            "LastUpdatedTime": "2023-03-20T16:11:08.513Z",
            "RollbackConfiguration": {},
            "StackStatus": "CREATE_COMPLETE",
            "DisableRollback": false,
            "NotificationARNs": [],
            "Capabilities": [
                "CAPABILITY_NAMED_IAM",
                "CAPABILITY_AUTO_EXPAND"
            ],
            "Outputs": [
                {
                    "OutputKey": "HeadNodeIp",
                    "OutputValue": "...",
                    "Description": "The Public IP address of the HeadNode"
                },
                {
                    "OutputKey": "ValidationMessages",
                    "OutputValue": "[]",
                    "Description": "Any warnings from cluster create or update operations."
                }
            ],
            "Tags": [],
            "EnableTerminationProtection": false,
            "DriftInformation": {
                "StackDriftStatus": "NOT_CHECKED"
            }
        }
    ]
}

Deleted temp working directory ...
```

charlesg3 committed:
`[cloudformation/custom_resource]` More pythonic flatten.

eantonin commented on cloudformation/tests/test_cluster_custom_resource.py:
nitpick, non-blocking: is this commit message wrong? Another `improve output, parametrize, deletion policy` ? Or is the commit view that is confusing me?

charlesg3 replied:
Hmm… that was the modification to that commit. I might have rebased incorrectly?

eantonin reviewed (approved):
LGTM
